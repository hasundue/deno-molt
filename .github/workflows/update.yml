name: Update

on:
  schedule:
    - cron: '0 0 * * *'
  workflow_dispatch:

permissions:
  contents: write
  pull-requests: write

jobs:
  update:
    name: Update
    runs-on: ubuntu-latest

    steps:
      - name: Checkout
        uses: actions/checkout@v4

      - name: Configure Git
        run: |
          git config --global user.name '${{ github.actor }}'
          git config --global user.email '${{ github.actor }}@users.noreply.github.com'

      - name: Run Molt
        run: deno task update --commit --pre-commit test --summary title.txt --report body.txt

      - name: Push changes
        run: git push

      - name: Create Pull Request
        run: |
          gh pr create --title "$(cat title.txt)" --body "$(cat body.txt)"
