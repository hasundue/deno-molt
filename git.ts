// Copyright 2023 Chiezo (Shun Ueda). All rights reserved. MIT license.

/**
 * A sub module of molt for git operations.
 *
 * ### Example
 *
 * To update all dependencies in a module and commit the changes to git:
 *
 * ```ts
 * import { DependencyUpdate } from "https://deno.land/x/molt@{VERSION}/mod.ts";
 * import { commitAll } from "https://deno.land/x/molt@{VERSION}/git.ts";
 *
 * const updates = await DependencyUpdate.collect("./mod.ts");
 *
 * commitAll(updates, {
 *   groupBy: (dependency) => dependency.name,
 *   composeCommitMessage: ({ group, version }) =>
 *     `build(deps): bump ${group} to ${version!.to}`,
 * });
 * ```
 *
 * @module
 */

export * from "./lib/git.ts";
