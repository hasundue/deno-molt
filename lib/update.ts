import { distinct } from "./std/collections.ts";
import { fromFileUrl } from "./std/path.ts";
import {
  createGraph,
  type CreateGraphOptions,
  init as initDenoGraph,
  load as defaultLoad,
  type ModuleJson,
} from "./x/deno_graph.ts";
import { findFileUp, toPath, toUrl } from "./path.ts";
import {
  ImportMap,
  ImportMapResolveResult,
  readImportMapJson,
  tryReadFromJson,
} from "./import_map.ts";
import {
  type Dependency,
  parse,
  resolveLatestVersion,
  toUrl as dependencyToUrl,
  type UpdatedDependency,
} from "./dependency.ts";

type DependencyJson = NonNullable<ModuleJson["dependencies"]>[number];

type If<T extends boolean, A, B> = T extends true ? A : B;

/**
 * Representation of an update to a dependency.
 */
export interface DependencyUpdate<IsMapped extends boolean = boolean> {
  /** Properties of the dependency being updated. */
  from: Dependency;
  /** Properties of the updated dependency. */
  to: UpdatedDependency;
  /**
   * The code of the dependency. Note that `type` in the DependencyJSON
   * is merged into `code` here for convenience.
   */
  code: {
    /** The original specifier of the dependency appeared in the code. */
    specifier: string;
    span: If<IsMapped, undefined, NonNullable<DependencyJson["code"]>["span"]>;
  };
  /** The full path to the module that imports the dependency.
   * @example "/path/to/mod.ts" */
  referrer: string;
  /** Information about the import map used to resolve the dependency. */
  map: If<
    IsMapped,
    {
      /** The full path to the import map used to resolve the dependency.
       * @example "/path/to/import_map.json" */
      source: string;
    } & ImportMapResolveResult<true>,
    undefined
  >;
}

class DenoGraph {
  static #initialized = false;

  static async ensureInit() {
    if (this.#initialized) {
      return;
    }
    await initDenoGraph();
    this.#initialized = true;
  }
}

export interface CollectOptions {
  /**
   * The working directory to resolve relative paths.
   * If not specified, the current working directory is used.
   * At present, this option is only used to find the import map.
   *
   * @example "/path/to/project"
   */
  cwd?: string | URL;
  /**
   * The path to the import map used to resolve dependencies.
   * If not specified, molt will automatically find deno.json or deno.jsonc
   * in the current working directory or parent directories.
   *
   * @example
   * ```ts
   * const updates = await DependencyUpdate.collect("mod.ts", {
   *   importMap: "import_map.json"
   *   // -> Use import_map.json in the current directory
   * });
   * ```
   */
  importMap?: string | URL;
  /**
   * A function to filter out dependencies.
   *
   * @example
   * ```ts
   * const updates = await DependencyUpdate.collect("mod.ts", {
   *   ignore: (dep) => dep.name === "deno.land/std"
   *   // -> Ignore all dependencies from deno.land/std
   * });
   * ```
   */
  ignore?: (dependency: Dependency) => boolean;
  /**
   * A function to pick dependencies.
   *
   * @example
   * ```ts
   * const updates = await DependencyUpdate.collect("mod.ts", {
   *   only: (dep) => dep.name === "deno.land/std"
   *   // -> Only pick dependencies from deno.land/std
   * });
   * ```
   */
  only?: (dependency: Dependency) => boolean;
}

/**
 * Collect dependencies from the given module(s) or Deno configuration file(s).
 * Local submodules are also checked recursively.
 *
 * @param from - The path(s) to the file(s) to collect dependencies from.
 * @param options - Options to customize the behavior.
 * @returns The list of dependencies.
 *
 * @example
 * ```ts
 * collect("mod.ts")
 * // -> Collect dependencies from mod.ts and its local submodules.
 * ```
 * @example
 * ```ts
 * collect("deno.json")
 * // -> Collect dependencies from the import map specified in deno.json
 * ```
 */
export async function collect(
  from: string | URL | (string | URL)[],
  options: CollectOptions = {},
): Promise<DependencyUpdate[]> {
  const froms = [from].flat();
  const urls = froms.map((path) => toUrl(path));

  const importMapPath = options.importMap ??
    await findFileUp(options.cwd ?? Deno.cwd(), "deno.json", "deno.jsonc");

  const importMap = importMapPath
    ? await tryReadFromJson(toUrl(importMapPath))
    : undefined;

  await DenoGraph.ensureInit();
  const graph = await createGraph(urls, {
    load,
    resolve: importMap?.resolveInner,
  });

  const updates: DependencyUpdate[] = [];
  await Promise.all([
    ...graph.modules
      .filter((m) => m.kind === "esm")
      .flatMap((m) =>
        m.dependencies?.map(async (dependency) => {
          const update = await _createDependencyUpdate(
            dependency,
            m.specifier,
            { ...options, importMap },
          );
          return update ? updates.push(update) : undefined;
        })
      ),
    ...graph.modules
      .filter((m) => m.kind === "asserted" && m.mediaType === "Json")
      .map(async (m) => {
        const results = await _collectFromImportMap(m.specifier, options);
        updates.push(...results);
      }),
  ]);
  return updates.sort((a, b) => a.to.name.localeCompare(b.to.name));
}

const load: NonNullable<CreateGraphOptions["load"]> = async (
  specifier,
) => {
  const url = new URL(specifier); // should not throw
  switch (url.protocol) {
    case "node:":
    case "npm:":
    case "jsr:":
      return {
        kind: "external",
        specifier,
      };
    case "http:":
    case "https:":
      return {
        kind: "external",
        specifier,
      };
    case "file:":
      return await defaultLoad(specifier);
    default:
      throw new Error(`Unsupported protocol: ${url.protocol}`);
  }
};

/**
 * Create a DependencyUpdate from the given dependency.
 * @param dependencyJson - The dependency to create an update from.
 * @param referrer - The URL of the module that imports the dependency.
 * @param options - Options to customize the behavior.
 * @returns The created DependencyUpdate.
 */
async function _createDependencyUpdate(
  dependencyJson: DependencyJson,
  referrer: string,
  options?: Pick<CollectOptions, "ignore" | "only"> & {
    importMap?: ImportMap;
  },
): Promise<DependencyUpdate | undefined> {
  const specifier = dependencyJson.code?.specifier ??
    dependencyJson.type?.specifier;
  if (!specifier) {
    throw new Error(
      `The dependency ${dependencyJson.specifier} in ${
        fromFileUrl(referrer)
      } has no resolved specifier.`,
      { cause: dependencyJson },
    );
  }
  const mapped = options?.importMap?.resolve(
    dependencyJson.specifier,
    referrer,
  ) as ImportMapResolveResult<true> | undefined;
  const dependency = parse(new URL(mapped?.value ?? specifier));
  if (options?.ignore?.(dependency)) {
    return;
  }
  if (options?.only && !options.only(dependency)) {
    return;
  }
  const latest = await resolveLatestVersion(dependency);
  if (!latest || latest.version === dependency.version) {
    return;
  }
  const span = dependencyJson.code?.span ?? dependencyJson.type?.span;
  if (!span) {
    throw new Error(
      `The dependency ${dependencyJson.specifier} in ${
        fromFileUrl(referrer)
      } has no span.`,
    );
  }
  return {
    from: dependency,
    to: latest,
    code: {
      // We prefer to put the original specifier here.
      specifier: dependencyJson.specifier,
      span,
    },
    referrer: toPath(referrer),
    map: mapped
      ? {
        source: options!.importMap!.path,
        ...mapped,
      }
      : undefined,
  };
}

async function _collectFromImportMap(
  specifier: ModuleJson["specifier"],
  options: Pick<CollectOptions, "ignore" | "only">,
): Promise<DependencyUpdate[]> {
  const json = await readImportMapJson(new URL(specifier));
  const updates: DependencyUpdate[] = [];
  await Promise.all(
    Object.entries(json.imports).map(
      async ([key, value]): Promise<DependencyUpdate | undefined> => {
        if (!URL.canParse(value)) {
          return;
        }
        const dependency = parse(new URL(value));
        if (options.ignore?.(dependency)) {
          return;
        }
        if (options.only && !options.only(dependency)) {
          return;
        }
        const latest = await resolveLatestVersion(dependency);
        if (!latest || latest.version === dependency.version) {
          return;
        }
        updates.push({
          from: dependency,
          to: latest,
          code: {
            specifier: value,
            span: undefined,
          },
          referrer: toPath(specifier),
          map: {
            source: toPath(specifier),
            resolved: value,
            key,
            value: dependencyToUrl(latest),
          },
        });
      },
    ),
  );
  return updates;
}

export type VersionChange = {
  from?: string;
  to: string;
};

export function getVersionChange(
  dependencies: DependencyUpdate[],
): VersionChange | undefined {
  const modules = distinct(dependencies.map((d) => d.to.name));
  if (modules.length > 1) {
    // Cannot provide a well-defined version prop
    return;
  }
  const tos = distinct(dependencies.map((d) => d.to.version));
  if (tos.length > 1) {
    throw new Error(
      "Multiple target versions are specified for a single module",
    );
  }
  const froms = distinct(dependencies.map((d) => d.from.version));
  return {
    from: froms.length === 1 ? froms[0] : undefined,
    to: tos[0],
  };
}
