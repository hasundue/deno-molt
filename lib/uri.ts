import { isAbsolute, relative, resolve, toFileUrl } from "./std/path.ts";
import { Brand } from "./types.ts";

export type DefaultProtocol<Scheme extends string> = Scheme extends
  "http" | "https" ? `${Scheme}://`
  : Scheme extends "file" ? `${Scheme}:///`
  : `${Scheme}:`;

export type URI<
  Scheme extends string = string,
  Protocol extends string = DefaultProtocol<Scheme>,
> = `${Protocol}${string}`;

export type RelativePath = Brand<string, "RelativePath">;
export type AbsolutePath = Brand<string, "AbsolutePath">;

export const URI = {
  from(path: string): URI<"file"> {
    let url: URL;
    try {
      url = new URL(path);
    } catch {
      return toFileUrl(
        isAbsolute(path) ? path : resolve(path),
      ).href as URI<"file">;
    }
    if (url.protocol !== "file:") {
      throw new TypeError();
    }
    return url.href as URI<"file">;
  },
  relative(uri: URI<"file">): RelativePath {
    return relative(Deno.cwd(), new URL(uri).pathname) as RelativePath;
  },
  absolute(uri: URI<"file">): AbsolutePath {
    return resolve(new URL(uri).pathname) as AbsolutePath;
  },
  ensure<S extends string>(...schemes: S[]): (uri: string) => URI<S> {
    return (uri) => {
      let url: URL;
      try {
        url = new URL(uri);
      } catch {
        throw new TypeError(`Invalid URI: ${uri}`);
      }
      if (schemes.includes(url.protocol.split(":")[0] as S)) {
        return uri as URI<S>;
      }
      throw new TypeError(`Unexpected URI protocol: ${url.protocol}`);
    };
  },
  is<S extends string>(uri: string, scheme: S): uri is URI<S> {
    try {
      const url = new URL(uri);
      return url.protocol.split(":")[0] === scheme;
    } catch {
      return false;
    }
  },
};
