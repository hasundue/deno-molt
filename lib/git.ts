import { distinct, mapN } from "./std/collections.ts";
import { relative } from "./std/path.ts";
import {
  associateByFile,
  type FileUpdate,
  writeFileUpdate,
  WriteOptions,
} from "./file.ts";
import { LockPart } from "./lockfile.ts";
import { CollectResult, DependencyUpdate } from "./update.ts";

export interface CommitProps {
  /** The name of the module group */
  group: string;
  version?: VersionChange;
}

export interface GitCommit extends CommitProps {
  message: string;
  updates: DependencyUpdate[];
  locks: LockPart[];
}

export interface CommitSequence {
  commits: GitCommit[];
  options: CommitOptions;
}

export interface CommitOptions extends WriteOptions {
  groupBy?: (dependency: DependencyUpdate) => string;
  composeCommitMessage?: (props: CommitProps) => string;
  preCommit?: (commit: GitCommit) => void | Promise<void>;
  postCommit?: (commit: GitCommit) => void | Promise<void>;
  gitAddOptions?: string[];
  gitCommitOptions?: string[];
}

const defaultCommitOptions = {
  groupBy: () => "dependencies",
  composeCommitMessage: ({ group, version }) => {
    let message = `build(deps): update ${group}`;
    if (version?.from) {
      message += ` from ${version.from}`;
    }
    if (version?.to) {
      message += ` to ${version.to}`;
    }
    return message;
  },
  gitAddOptions: [],
  gitCommitOptions: [],
} satisfies CommitOptions;

/**
 * Write the given `CollectResult` to file system.
 * @returns A promise that resolves when all updates are committed.
 */
export function commit(
  result: CollectResult,
  options?: CommitOptions,
) {
  return execute(createCommitSequence(result, {
    ...defaultCommitOptions,
    ...options,
  }));
}

/**
 * Create a sequence of commits from the given `CollectResult`.
 */
export function createCommitSequence(
  result: CollectResult,
  options?: CommitOptions,
): CommitSequence {
  const _options = { ...defaultCommitOptions, ...options };
  const groups = new Map<string, DependencyUpdate[]>();
  for (const u of result.updates) {
    const key = _options.groupBy(u);
    if (!groups.has(key)) {
      groups.set(key, []);
    }
    groups.get(key)!.push(u);
  }
  const commits: GitCommit[] = Array.from(groups.entries()).map((
    [group, updates],
  ) => {
    const version = getVersionChange(updates);
    return ({
      group,
      version,
      message: _options.composeCommitMessage({ group, version }),
      updates,
      locks: result.locks,
    });
  }).sort((a, b) => a.group.localeCompare(b.group));
  return { commits, options: _options };
}

export function execute(
  commit: GitCommit,
  options?: CommitOptions,
): Promise<void>;
export function execute(sequence: CommitSequence): Promise<void>;

export function execute(
  commit: GitCommit | CommitSequence,
  options?: CommitOptions,
): Promise<void> {
  if ("commits" in commit) {
    return execCommitSequence(commit);
  }
  return execCommit(commit, options);
}

async function execCommitSequence(
  sequence: CommitSequence,
) {
  for (const commit of sequence.commits) {
    await execCommit(commit, sequence.options);
  }
}

async function execCommit(
  commit: GitCommit,
  options?: CommitOptions,
) {
  const results = associateByFile(commit);
  await writeFileUpdate(results);
  await options?.preCommit?.(commit);
  await addCommand(results, options?.gitAddOptions ?? []);
  await commitCommand(commit.message, options?.gitCommitOptions ?? []);
  await options?.postCommit?.(commit);
}

async function addCommand(
  results: FileUpdate[],
  options: string[],
) {
  const files = results.map((result) => relative(Deno.cwd(), result.path));
  const { code, stderr } = await new Deno.Command("git", {
    args: ["add", ...options, ...files],
  }).output();
  if (code !== 0) {
    throw new Error(new TextDecoder().decode(stderr));
  }
}

async function commitCommand(
  message: string,
  options: string[],
) {
  const { code, stderr } = await new Deno.Command("git", {
    args: ["commit", ...options, "-m", message],
  }).output();
  if (code !== 0) {
    throw new Error(new TextDecoder().decode(stderr));
  }
}

export type VersionChange = {
  from?: string;
  to: string;
};

export function getVersionChange(
  dependencies: DependencyUpdate[],
): VersionChange | undefined {
  const modules = distinct(dependencies.map((d) => d.to.name));
  if (modules.length > 1) {
    // Cannot provide a well-defined version prop
    return;
  }
  const tos = distinct(dependencies.map((d) => d.to.version));
  if (tos.length > 1) {
    throw new Error(
      "Multiple target versions are specified for a single module",
    );
  }
  const froms = distinct(mapN(dependencies, (d) => d.from?.version));
  return {
    from: froms.length === 1 ? froms[0] : undefined,
    to: tos[0],
  };
}
