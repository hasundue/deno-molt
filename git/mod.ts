// Copyright 2023 Chiezo (Shun Ueda). All rights reserved. MIT license.

/**
 * A sub module of molt for git operations.
 *
 * ### Example
 *
 * To update all dependencies in a module and commit the changes to git:
 *
 * ```ts
 * import { DependencyUpdate } from "https://deno.land/x/molt@{VERSION}/mod.ts";
 * import { commitAll } from "https://deno.land/x/molt@{VERSION}/git/mod.ts";
 *
 * const updates = await DependencyUpdate.collect("./mod.ts");
 *
 * commitAll(updates, {
 *   groupBy: (dependency) => dependency.name,
 *   composeCommitMessage: ({ group, version }) =>
 *     `build(deps): bump ${group} to ${version!.to}`,
 * });
 * ```
 *
 * @module
 */

import { DependencyUpdate } from "../src/update.ts";
import { FileUpdate } from "../src/file.ts";
import { createVersionProp, type VersionProp } from "../src/versions.ts";
import { URI } from "../lib/uri.ts";

export interface CommitProps {
  /** The name of the module group */
  group: string;
  version?: VersionProp;
}

export interface CommitOptions {
  groupBy?: (dependency: DependencyUpdate) => string;
  composeCommitMessage?: (props: CommitProps) => string;
  gitAddOptions?: string[];
  gitCommitOptions?: string[];
}

export const defaultCommitOptions = {
  groupBy: () => "dependencies",
  composeCommitMessage: ({ group, version }) => {
    let message = `build(deps): update ${group}`;
    if (version?.from) {
      message += ` from ${version.from}`;
    }
    if (version?.to) {
      message += ` to ${version.to}`;
    }
    return message;
  },
  gitAddOptions: [],
  gitCommitOptions: [],
} satisfies CommitOptions;

export interface GitCommit extends CommitProps {
  message: string;
  updates: DependencyUpdate[];
}

export const GitCommit = {
  sequence: createGitCommitSequence,
  exec: execGitCommit,
  execAll: execGitCommitSequence,
};

export interface GitCommitSequence {
  commits: GitCommit[];
  options: CommitOptions;
}

export interface ExecGitCommitSequenceOptions {
  preCommit?: (commit: GitCommit) => void;
  postCommit?: (commit: GitCommit) => void;
}

export function commitAll(
  updates: DependencyUpdate[],
  options?: CommitOptions & ExecGitCommitSequenceOptions,
): void {
  execGitCommitSequence(
    createGitCommitSequence(updates, options),
    options,
  );
}

function createGitCommitSequence(
  updates: DependencyUpdate[],
  options?: Partial<CommitOptions>,
): GitCommitSequence {
  const _options = { ...defaultCommitOptions, ...options };
  const groups = new Map<string, DependencyUpdate[]>();
  for (const u of updates) {
    const key = _options.groupBy(u);
    if (!groups.has(key)) {
      groups.set(key, []);
    }
    groups.get(key)!.push(u);
  }
  const version = createVersionProp(updates);
  const commits: GitCommit[] = Array.from(groups.entries()).map((
    [group, updates],
  ) => ({
    group,
    version,
    message: _options.composeCommitMessage({ group, version }),
    updates,
  }));
  return { commits, options: _options };
}

function execGitCommitSequence(
  sequence: GitCommitSequence,
  options?: ExecGitCommitSequenceOptions,
) {
  for (const commit of sequence.commits) {
    options?.preCommit?.(commit);
    execGitCommit(commit, sequence.options);
    options?.postCommit?.(commit);
  }
}

function execGitCommit(
  commit: GitCommit,
  options?: CommitOptions,
) {
  const results = FileUpdate.collect(commit.updates);
  FileUpdate.writeAll(results);
  _add(results, options?.gitAddOptions ?? []);
  _commit(commit.message, options?.gitCommitOptions ?? []);
}

function _add(
  results: FileUpdate[],
  options: string[],
) {
  const files = results.map((result) => URI.relative(result.specifier));
  const command = new Deno.Command("git", {
    args: ["add", ...options, ...files],
  });
  const { code } = command.outputSync();
  if (code !== 0) {
    throw new Error(`git add failed: ${code}`);
  }
}

function _commit(
  message: string,
  options: string[],
) {
  const command = new Deno.Command("git", {
    args: ["commit", ...options, "-m", message],
  });
  const { code } = command.outputSync();
  if (code !== 0) {
    throw new Error(`git commit failed: ${code}`);
  }
}
