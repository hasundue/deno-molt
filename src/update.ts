import {
  createGraph,
  init as initDenoGraph,
  type ModuleJson,
} from "../lib/x/deno_graph.ts";
import { RelativePath, URI } from "../lib/uri.ts";
import type { SemVerString } from "./types.ts";
import { ImportMap, ImportMapJson } from "./import_map.ts";
import { Dependency } from "./dependency.ts";
import { load } from "./loader.ts";

type DependencyJson = NonNullable<ModuleJson["dependencies"]>[number];

/** Representation of a dependency update. */
export interface DependencyUpdate extends Omit<Dependency, "version"> {
  /** The fully resolved specifier of the dependency. */
  specifier: URI<"http" | "https" | "npm">;
  version: {
    from: SemVerString;
    to: SemVerString;
  };
  /** The code of the dependency. */
  code?: {
    /** The original specifier of the dependency appeared in the code. */
    specifier: string;
    span: NonNullable<DependencyJson["code"]>["span"];
  };
  type?: {
    specifier: string;
    span: NonNullable<DependencyJson["type"]>["span"];
  };
  /** The specifier of the module that imports the dependency. */
  referrer: URI<"file">;
  /** Information about the import map used to resolve the dependency. */
  map?: {
    /** The path to the import map used to resolve the dependency. */
    source: URI<"file">;
    from: string;
    /** The string in the dependency specifier being replaced by the import map.
     * Mapping on a file specifier should not happen. */
    to: URI<"http" | "https" | "npm">;
  };
}

export const DependencyUpdate = {
  collect,
  applyToModule,
  applyToImportMap,
  withRelativePath,
};

class DenoGraph {
  static #initialized = false;

  static async ensureInit() {
    if (this.#initialized) {
      return;
    }
    await initDenoGraph();
    this.#initialized = true;
  }
}

export async function collect(
  entrypoints: string | string[],
  options: {
    importMap?: string;
  } = {},
): Promise<DependencyUpdate[]> {
  // This could throw if the entrypoints are not valid URIs.
  const specifiers = [entrypoints].flat().map((path) => URI.from(path));

  // Ensure the deno_graph WASM module is initialized.
  await DenoGraph.ensureInit();

  const importMap = options.importMap
    ? await ImportMap.readFromJson(options.importMap)
    : undefined;

  const graph = await createGraph(specifiers, {
    load,
    resolve: importMap ? importMap.resolveSimple : undefined,
  });

  const updates: DependencyUpdate[] = [];
  await Promise.all(
    graph.modules.flatMap((module) =>
      module.dependencies?.map(async (dependency) => {
        const update = await _create(
          dependency,
          URI.from(module.specifier),
          { importMap },
        );
        return update ? updates.push(update) : undefined;
      })
    ),
  );
  return updates;
}

export async function _create(
  dependency: DependencyJson,
  referrer: URI<"file">,
  options?: {
    importMap?: ImportMap;
  },
): Promise<DependencyUpdate | undefined> {
  if (!dependency?.code?.specifier && !dependency?.type?.specifier) {
    throw new Error(
      `The dependency ${dependency.specifier} in ${
        URI.relative(referrer)
      } has no resolved specifier.`,
    );
  }
  const specifier = dependency.code?.specifier ?? dependency.type?.specifier!;
  const newSemVer = await Dependency.resolveLatestSemVer(new URL(specifier));
  if (!newSemVer) {
    return;
  }
  const props = Dependency.parseProps(new URL(specifier));
  if (!props) {
    return;
  }
  const mapped = options?.importMap?.resolve(
    dependency.specifier,
    referrer,
  );
  return {
    ...props,
    // We prefer to put the fully resolved specifier here.
    specifier: URI.ensure("http", "https", "npm")(specifier),
    code: dependency.code && {
      // We prefer to put the original specifier here.
      specifier: dependency.specifier,
      span: dependency.code.span,
    },
    type: dependency.type && {
      // We prefer to put the original specifier here.
      specifier: dependency.specifier,
      span: dependency.type.span,
    },
    version: {
      from: props.version as SemVerString,
      to: newSemVer as SemVerString,
    },
    referrer,
    map: mapped
      ? {
        source: options!.importMap!.specifier,
        from: mapped.from!,
        to: URI.ensure("http", "https", "npm")(mapped.to!),
      }
      : undefined,
  };
}

function applyToModule(
  /** The dependency update to apply. */
  update: DependencyUpdate,
  /** Content of the module to update. */
  content: string,
): string {
  const span = update.code?.span ?? update.type?.span!;
  if (span.start.line !== span.end.line) {
    throw new Error(
      `The import specifier ${update.specifier} in ${update.referrer} is not a single line`,
    );
  }
  const line = span.start.line;
  const lines = content.split("\n");

  lines[line] = lines[line].slice(0, span.start.character) +
    `"${
      update.specifier.replace(
        update.version.from,
        update.version.to,
      )
    }"` +
    lines[line].slice(span.end.character);

  return lines.join("\n");
}

export function applyToImportMap(
  /** The dependency update to apply. */
  update: DependencyUpdate,
  /** Content of the import map to update. */
  content: string,
): string {
  const json = JSON.parse(content) as ImportMapJson;
  json.imports[update.map!.from] = update.map!.to.replace(
    update.version.from,
    update.version.to,
  );
  return JSON.stringify(json, null, 2);
}

export type DependencyUpdateWithRelativePath =
  & Omit<DependencyUpdate, "specifier" | "map">
  & {
    specifier: RelativePath;
    map?: {
      source: RelativePath | URI<"http" | "https" | "npm">;
      from: string;
      to: URI<"http" | "https" | "npm">;
    };
  };

/**
 * Convert specifiers in the dependency update to relative paths for subsequent operations.
 */
function withRelativePath(
  update: DependencyUpdate,
): DependencyUpdateWithRelativePath {
  return {
    ...update,
    specifier: _relativeIfFile(update.specifier),
    map: update.map && {
      ...update.map,
      source: URI.relative(update.map.source),
    },
  } as DependencyUpdateWithRelativePath;
}

function _relativeIfFile(
  specifier: URI<"file" | "http" | "https" | "npm">,
) {
  return URI.is(specifier, "file") ? URI.relative(specifier) : specifier;
}
