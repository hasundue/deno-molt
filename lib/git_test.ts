import {
  afterAll,
  assertSnapshot,
  assertSpyCalls,
  beforeAll,
  beforeEach,
  describe,
  it,
} from "./std/testing.ts";
import {
  assertFindSpyCall,
  createCommandStub,
  FileSystemFake,
  ReadTextFileStub,
  WriteTextFileStub,
} from "./testing.ts";
import { URI } from "./uri.ts";
import { DependencyUpdate } from "./update.ts";
import { commitAll } from "./git.ts";

describe("commitAll()", () => {
  let updates: DependencyUpdate[];
  let fileSystemFake: FileSystemFake;
  let writeTextFileStub: WriteTextFileStub;
  let readTextFileStub: ReadTextFileStub;
  let CommandStub: ReturnType<typeof createCommandStub>;

  beforeAll(async () => {
    updates = await DependencyUpdate.collect(
      "./test/fixtures/direct-import/mod.ts",
    );
    fileSystemFake = new FileSystemFake();
    readTextFileStub = ReadTextFileStub.create(fileSystemFake, {
      readThrough: true,
    });
    writeTextFileStub = WriteTextFileStub.create(fileSystemFake);
  });

  afterAll(() => {
    writeTextFileStub.restore();
    readTextFileStub.restore();
    Deno.Command = CommandStub.original;
  });

  beforeEach(() => {
    fileSystemFake.clear();
    CommandStub = createCommandStub();
    Deno.Command = CommandStub;
  });

  // "git add src/fixtures/mod.ts src/fixtures/lib.ts",
  it("no grouping", async (t) => {
    await commitAll(updates);
    // TODO: Can't test this because of the order of targets is not guaranteed.
    // assertGitAdd(CommandStub, "src/fixtures/mod.ts", "src/fixtures/lib.ts");
    assertGitCommit(CommandStub, "build(deps): update dependencies");
    assertSpyCalls(CommandStub, 2);
    await assertSnapshot(t, Array.from(fileSystemFake.values()));
  });

  it("group by dependency name", async (t) => {
    await commitAll(updates, {
      groupBy: (update) => update.name,
      composeCommitMessage: ({ group }) => `build(deps): update ${group}`,
    });
    assertGitAdd(CommandStub, "test/fixtures/direct-import/mod.ts");
    assertGitCommit(CommandStub, "build(deps): update node-emoji");
    assertGitAdd(CommandStub, "test/fixtures/direct-import/mod.ts");
    assertGitCommit(CommandStub, "build(deps): update deno.land/x/deno_graph");
    // TODO: Can't test this because of the order of targets is not guaranteed.
    // assertGitAdd(CommandStub, "src/fixtures/lib.ts", "src/fixtures/mod.ts");
    assertGitCommit(CommandStub, "build(deps): update deno.land/std");
    assertSpyCalls(CommandStub, 6);
    await assertSnapshot(t, Array.from(fileSystemFake.values()));
  });

  it("group by module (file) name", async (t) => {
    await commitAll(updates, {
      groupBy: (update) => update.referrer,
      composeCommitMessage: ({ group }) => {
        const uri = URI.ensure("file")(group);
        const relative = URI.relative(uri);
        return `build(deps): update ${relative}`;
      },
    });
    assertGitAdd(CommandStub, "test/fixtures/direct-import/mod.ts");
    assertGitCommit(
      CommandStub,
      "build(deps): update test/fixtures/direct-import/mod.ts",
    );
    assertGitAdd(CommandStub, "test/fixtures/direct-import/lib.ts");
    assertGitCommit(
      CommandStub,
      "build(deps): update test/fixtures/direct-import/lib.ts",
    );
    assertSpyCalls(CommandStub, 4);
    await assertSnapshot(t, Array.from(fileSystemFake.values()));
  });
});

function normalizePath(path: string) {
  return Deno.build.os === "windows" ? path.replaceAll("/", "\\") : path;
}

function assertGitCommit(
  Command: ReturnType<typeof createCommandStub>,
  message: string,
) {
  assertFindSpyCall(Command, {
    args: [
      "git",
      { args: ["commit", "-m", message] },
    ],
  });
}

function assertGitAdd(
  Command: ReturnType<typeof createCommandStub>,
  ...paths: string[]
) {
  assertFindSpyCall(Command, {
    args: [
      "git",
      { args: ["add", ...paths.map(normalizePath)] },
    ],
  });
}
