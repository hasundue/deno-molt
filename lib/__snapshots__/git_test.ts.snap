export const snapshot = {};

snapshot[`commitAll() > no grouping 1`] = `
[
  'import { VERSION } from "https://deno.land/std@0.204.0/version.ts";
import { createGraph } from "https://deno.land/x/deno_graph@0.59.1/mod.ts";
import { emojify } from "npm:node-emoji@2.1.0";
import { noop } from "./lib.ts";
',
  'import { VERSION } from "https://deno.land/std@0.204.0/version.ts";
export const noop = () => {};
',
]
`;

snapshot[`commitAll() > group by dependency name 1`] = `
[
  'import { VERSION } from "https://deno.land/std@0.200.0/version.ts";
import { createGraph } from "https://deno.land/x/deno_graph@0.50.0/mod.ts";
import { emojify } from "npm:node-emoji@2.1.0";
import { noop } from "./lib.ts";
',
  'import { VERSION } from "https://deno.land/std@0.204.0/version.ts";
export const noop = () => {};
',
]
`;

snapshot[`commitAll() > group by module (file) name 1`] = `
[
  'import { VERSION } from "https://deno.land/std@0.204.0/version.ts";
import { createGraph } from "https://deno.land/x/deno_graph@0.59.1/mod.ts";
import { emojify } from "npm:node-emoji@2.1.0";
import { noop } from "./lib.ts";
',
  'import { VERSION } from "https://deno.land/std@0.204.0/version.ts";
export const noop = () => {};
',
]
`;
